[hail] Add table and matrix table indices

chrisvittal opened the pull request:
This currently is exposed in `read_table` and `read_matrix_table` as
undocumented optional parameters `_intervals` and `_filter_intervals`
which takes a list of python `Interval`s that are used either as
a filter or a repartition.

This adds an `IndexedRVDSpec` as the primary container format for
indexed data, and increments the file version to 1.1.0.

One index file is written per partition. For matrix tables, the offset
to a particular key for the entries is stored in the `entries_offset`
field of the annotation that an index may contain. We use the new
`IndexSpec` to retrive the appropriate offset from the index so that we
can seek to the proper offset in the partition.

When writing data with a blocked spec (like the default) we use virtual
index offsets similar to tabix. The high 48 bits are used to indicate
the file offset to the start of a block, and the low 16 bits are used to
indicate the offset from the start of the (possibly decompressed) block.

cc @cseed

tpoterba commented on hail/python/hail/ir/matrix_reader.py:
```diff
@@ -20,18 +22,46 @@ def __eq__(self, other):
+                      filter_intervals=bool)
+    def __init__(self, path, intervals, filter_intervals):
+        if intervals is not None:
+            t = hl.expr.impute_type(intervals)
```
I think this logic should be in read_matrix_table -- we generally assume that parameters on the IR have been validated.

tpoterba replied:
also, shouldn't there be a check that the interval type agrees with the key type?

chrisvittal replied:
Can't do that until this is wrapped in a `Table`/`MatrixTable` (python classes). We have no access to the key type at this point. I can certainly move much of the logic into `read(_matrix)_table`.

tpoterba replied:
Yeah, realized that after I commented and was looking at some of the code i master. I think it would be preferable to do the type lookup in read_matrix_table and then pass it into the IR node, but that's definitely out of scope of this change.

tpoterba commented on hail/python/hail/ir/matrix_reader.py:
```diff
@@ -20,18 +22,46 @@ def __eq__(self, other):
+                self._interval_type = hl.tarray(hl.tinterval(hl.tstruct(__point=pt)))
+
         self.path = path
+        self.intervals = hl.eval(intervals) if intervals is not None else None
```
yeah, we definitely can't call `hl.eval()` in IR initialization. If we want to recons the IR, suddenly that's a super slow operation.

tpoterba commented on hail/python/hail/ir/table_reader.py:
```diff
@@ -17,18 +19,46 @@ def __eq__(self, other):
+    @typecheck_method(path=str,
+                      intervals=nullable(sequenceof(anytype)),
+                      filter_intervals=bool)
+    def __init__(self, path, intervals, filter_intervals):
```
same as above.

tpoterba commented on hail/python/hail/methods/impex.py:
```diff
@@ -1710,9 +1710,12 @@ def import_plink(bed, bim, fam,
            _drop_cols=bool,
            _drop_rows=bool)
-def read_matrix_table(path, _drop_cols=False, _drop_rows=False) -> MatrixTable:
+def read_matrix_table(path, intervals=None, filter_intervals=False, _drop_cols=False,
```
let's add a `*` between `path` and the rest of the args, so that they become keyword-only args and we can change the order without affecting backwards compatibility.

tpoterba replied:
also, if these params are to be public, they need documentation.

tpoterba commented on hail/python/hail/methods/impex.py:
```diff
@@ -2107,8 +2111,10 @@ def index_bgen(path,
-@typecheck(path=str)
-def read_table(path) -> Table:
+@typecheck(path=str,
+           intervals=nullable(sequenceof(anytype)),
```
same as above.

tpoterba commented on hail/python/hail/methods/misc.py:
```diff
@@ -91,7 +91,7 @@ def maximal_independent_set(i, j, keep=True, tie_breaker=None, keyed=True) -> Ta
 
     When multiple nodes have the same degree, this algorithm will order the
     nodes according to ``tie_breaker`` and remove the *largest* node.
-    
```
whitespace change, and not a fix to something against default formatting?

chrisvittal replied:
My editor does this automatically. I can revert this in particular if we want, but also I feel that it's good style to not have trailing whitespace.

tpoterba replied:
yeah, I'm sympathetic, but it does make the diff harder too read. A setting that applies the transformation only in lines otherwise modified would be great -- does that exist?

tpoterba commented on hail/src/main/scala/is/hail/HailContext.scala:
```diff
@@ -528,6 +613,50 @@ class HailContext private(
+    val nPartitions = partFiles.length
+    val localHadoopConfBc = hadoopConfBc
+    val idxMetadata1 = IndexReader.readMetadata(
+      hadoopConf, s"$path/$idxPath/${ partFiles.head }.idx")
```
this feels wrong. The index metadata should be in the `Spec`, I think.

chrisvittal replied:
I agree. However, that's a future enhancement that probably requires redoing how we handle indices in a backwards incompatible way that would break BGEN indices.

tpoterba replied:
see my other comment -- I now think it should probably be in the RVDSpec.

tpoterba replied:
that seems like the right level for it

tpoterba commented on hail/src/main/scala/is/hail/expr/ir/MatrixValue.scala:
```diff
@@ -68,6 +68,7 @@ case class MatrixValue(
     val colsSpec = TableSpec(
       FileFormat.version.rep,
       is.hail.HAIL_PRETTY_VERSION,
+      None,  // cols are not indexed
```
should/could they be?

tpoterba commented on hail/src/main/scala/is/hail/table/Table.scala:
```diff
@@ -27,22 +28,65 @@ case object Descending extends SortOrder
+    filterIntervals: Boolean
+  ): RVD = {
+    if (index_spec.isEmpty)
+      fatal(s"attempted to read with index from an unindexed table: $path")
```
A user will hit this message if attempting to use the `intervals` param with an old table, right? Maybe we should add a second line recommending a read/write?

chrisvittal replied:
A user will probably hit the one in `TableRead` before they hit this one. But better error messages definitely need to happen.

tpoterba commented on hail/src/main/scala/is/hail/table/Table.scala:
```diff
@@ -27,22 +28,65 @@ case object Descending extends SortOrder
+      fatal(s"attempted to read with index from an unindexed table: $path")
+
+    val requestedKey = table_type.key.takeWhile(rowType.hasField)
+    val rows = rowsSpec(path).asInstanceOf[OrderedRVDSpec]
```
I don't think you can do this cast

chrisvittal replied:
I can definitely do this cast. All `1.1.0` table's `rowsSpec`s are `OrderedRVDSpec`s. I need access to the partitioner.

tpoterba replied:
Won't casting to an AbstractRVDSpec also work? I do realize this will work right now, but that seems safer (if for instance, we upgrade the file format without changing the index stuff)

tpoterba commented on hail/src/main/scala/is/hail/variant/MatrixTable.scala:
```diff
@@ -140,7 +142,9 @@ case class MatrixTableSpec(
-  val version: SemanticVersion = SemanticVersion(1, 0, 0)
+  val version: SemanticVersion = SemanticVersion(1, 1, 0)
+
+  val indicesVersion: SemanticVersion = SemanticVersion(1, 1, 0)
```
adding *another* version to worry about seems worse than incrementing the minor version. I will push strongly for the latter approach.

chrisvittal replied:
What happens when we increment again? How do we make sure that `1.1.0` tables are still read appropriately?

The intent of `indicesVersion` (willing to bikeshed on name) is to enshrine in a constant the version in the `1.0` series of file formats where indices appeared. It is used to compare against the TableSpec's version in TableRead.

tpoterba replied:
> How do we make sure that 1.1.0 tables are still read appropriately

You mean, what happens when we update the index design? That's a good point, we need to add an IndexSpec that contains the relevant information.

Though now that I'm think about this... Can't the index metadata just be part of the RVDSpec?

chrisvittal replied:
I don't like that because the RVDSpec is not versioned.

tpoterba replied:
it is -- the existing concrete specs are immutable, you can't change OrderedRVDSpec for instance. You could add an IndexedRVDSpec, however, that contains the index metadata!

chrisvittal commented:
I ran the python tests with my latest changes and they all passed. I'll work on addressing comments and adding tests, but this should be ready for another round of code review.

chrisvittal commented:
Okay. This is 'done' (pending more review and revisions).

tpoterba commented:
I'll try to give this a once-over this afternoon.

tpoterba commented:
Also need to see benchmarks.

chrisvittal commented:
Still need to write benchmarks, but I wrote a new type to hold the options for the indexed reads, please take a look.

chrisvittal commented:
Ok, benchmarks are pretty bad, these are the range table benchmarks from #6529

#### No Index
```
running write_range_table_p1000...
    Mean, Median: 11.79s, 11.18s
running write_range_table_p100...
    Mean, Median: 4.76s, 4.67s
running write_range_table_p10...
    Mean, Median: 3.28s, 3.03s
```
#### Index
```
running write_range_table_p1000...
    Mean, Median: 28.60s, 28.88s
running write_range_table_p100...
    Mean, Median: 10.17s, 10.20s
running write_range_table_p10...
    Mean, Median: 9.52s, 9.44s
```

cseed commented:
For where we are, this seems pretty good.  Range table is the worst case because the key consists of all the data.  For the key, the index should have roughly 2x cost (stores an extra copy of the key, along with metadata like the file offset).  What's more, we're using Java values for the keys and annotations, which is also really bad.

A better to do would be have each node (internal or leaf) have a region that stores the keys, and directly write that out instead of using going from Java object to region value to serialization.

tpoterba commented:
> A better to do would be have each node (internal or leaf) have a region that stores the keys, and directly write that out instead of using going from Java object to region value to serialization.

Chris and I discussed this lasat week.

tpoterba commented:
Can we do a bit of profiling to figure out where it's spending the time?

chrisvittal commented:
It is spending it's time in `IndexWriter.+=`. For this particular example, we more than triple the data we're writing, before we were writing an `int32`, now we're writing an `int32` row, along with a B-Tree with an `int32` key and an `int64` offset field. So honestly this is pretty good. This was always going to make writes slower.

tpoterba commented:
> It is spending it's time in IndexWriter.+=

Great, this can also be fixed by not using Java objects. That makes me feel good about things.
